imxrt-hal/src/can/mod.rs

```diff
@@ -0,0 +1,972 @@
+//! Flexible Controller Area Network (FlexCAN)
+//!
+//! The FlexCAN module provides a CanBus peripheral that implements
+//! the `embedded_hal::can` traits.
+//!//!
+//! # Example
+//!
+//! ```no_run
+//! use imxrt_hal;
+//!
+//! let mut peripherals = imxrt_hal::Peripherals::take().unwrap();
+//!
+//! let (can1_builder, _) = peripherals.can.clock(
+//!     &mut peripherals.ccm.handle,
+//!     imxrt_hal::ccm::can::ClockSelect::OSC,
+//!     imxrt_hal::ccm::can::PrescalarSelect::DIVIDE_1,
+//! );
+//!
+//! let mut can1 = can1_builder.build(
+//!     peripherals.iomuxc.ad_b1.p08,
+//!     peripherals.iomuxc.ad_b1.p09
+//! );
+//!
+//! can1.set_baud_rate(1_000_000);
+//! can1.set_max_mailbox(16);
+//! can1.enable_fifo();
+//! can1.set_fifo_interrupt(true);
```

Comment:
Question about this API and behavior. I'm effectively testing this documentation example through the `can.rs` example proposed in the teensy4-rs repo (as of [this commit](https://github.com/mciantyre/teensy4-rs/pull/124/commits/d176a6d17ea2c7e022aad9d21461168d3cc59473)).

1. I'm changing this driver configuration to `false`, and keeping all other `can1` calls in the block the same.

    ```rust
    can1.set_fifo_interrupt(false);
    ```

2. I'm then unmasking the CAN1 interrupt before the LED-toggling-CAN-I/O loop.

    ```rust
    // NEW:
    unsafe { cortex_m::peripheral::NVIC::unmask(bsp::interrupt::CAN1) };
    loop { /* Delay, toggle LED, read CAN mailboxes, send CAN frame */ }
    ```

3. I implementing the CAN1 ISR elsewhere in the example. It looks like

    ```rust
    use bsp::interrupt;
    #[cortex_m_rt::interrupt]
    fn CAN1() {
        log::warn!("CAN1 interrupt fired");
    }
    ```

I'm expecting step 1 to disable the CAN interrupts; that seems to be the only public method with something implying interrupt activation in the name. So even though I'm unmasking the CAN1 interrupt, I'm never expecting the ISR to activate.

But when I actually build and run this, I see that the CAN1 ISR is firing repeatedly. I'm connecting to the USB serial console to see all the log warnings.

Is there something else in the driver that's enabling interrupts, and can it be changed by the user? Or, am I using this driver incorrectly?

Reply:
From the [manual](https://www.pjrc.com/teensy/IMXRT1060RM_rev2.pdf) (pg 2553): 

_The module can generate up to 70 interrupt sources (64 interrupts due to message buffers
and 6 interrupts due to Ored interrupts from MBs, Bus Off, Error, Tx Warning, Rx
Warning and Wake Up)).

If the Rx FIFO is enabled (bit RFEN on MCR set), the interrupts corresponding to MBs 0
to 7 have a different behavior. Bit 7 of the IFLAG1 becomes the "FIFO Overflow" flag;
bit 6 becomes the FIFO Warning flag, bit 5 becomes the "Frames Available in FIFO flag"
and bits 4-0 are unused. See Interrupt Flags 1 Register (FLEXCAN_IFLAG1) for more
information.

A combined interrupt for all MBs is also generated by an Or of all the interrupt sources
from MBs. This interrupt gets generated when any of the Mailboxes or FIFO generates an
interrupt. The Arm must read the IFLAG Registers to determine which MB or FIFO
caused the interrupt.

The other 5 interrupt sources (Bus Off, Error, Tx Warning, Rx Warning and Wake Up
generate interrupts like the MB ones, and can be read from both the Error and Status
Register 1 and 2. The Bus Off, Error, Tx Warning and Rx Warning interrupt mask bits
are located in the Control 1 Register and the Wake-Up interrupt mask bit is located in the
MCR._

This method is currently only setting the BUF5I bit in IMASK1 register to interrupt when a frame is available (Frames Available in FIFO flag...):

_If the Rx FIFO is not enabled, this bit flags the interrupt for MB5. If the Rx FIFO is enabled, this flag
indicates that at least one frame is available to be read from the Rx FIFO.
This flag is cleared by the FlexCAN whenever the bit MCR[RFEN] is changed by Arm writes_

I agree this could be more transparent. More importantly, masking the interrupt vector for the interrupt reason would be beneficial. 

This would let us debug why the interrupt is always firing for you (likely Bus Off with no transceiver attached.)

Reply:
When I run the `can.rs` example in the teensy4-bsp project, it logs a few control registers before entering its I/O loop. Here's three register values I see, followed by a discussion of interrupt-specific register fields.

```
[INFO imxrt_hal::can]: MCR: 64EB300F
```

In `MCR`,

- `WRNEN` is set. However, we're never setting `CTRL1[TWRNMSK]` or `CTRL1[RWRNMSK]`, so (from what I can tell) we shouldn't observe an interrupt when error counters cross their threshold. If that's correct, I'm not sure we need to set `WRNEN`.
- `WAKMSK` and `SLFWAK` are set. This means that we could trigger an interrupt when exiting doze or stop modes. (This definitely isn't happening in my limited testing, but it may be a non-obvious behavior of this driver.)

```
[INFO imxrt_hal::can]: CTRL1: 1624002
```

In `CTRL1`,

- `BOFFMSK` is clear, so there's no "bus off" interrupt.
- `ERRMSK` is set, so we generate an interrupt when `ESR1[ERRINT]` is high. This is a hasty observation, but a project-wide CTRL+F for "[ESR1](https://docs.rs/imxrt-ral/0.4.4/imxrt_ral/imxrt106/peripherals/can/ESR1/index.html)" shows no hits. This could indicate we're not clearing error interrupts.
- `TWRNMSK` and `RWRNMSK` are both clear; see `MCR[WRNEN]` discussion, above.

```
[INFO imxrt_hal::can]: CTRL2: 870000
```

In `CTRL2`,

- `BOFFDONEMSK` is clear, so there's no "bus off done" interrupt.

I'm not evaluating any message buffer interrupts right now. Nevertheless, I'm wary of `ERRMSK` being set, especially if we're not handling any error and status flags.